README.md
# Axum Strangler

A utility crate to be able to easily use the Strangler Fig pattern with the Axum crate without having to use some sort of gateway.

## Goal of the crate

To support the usecase where you want to rewrite services in Rust, but you can't justify the
cost of migrating everything over all at once.
With the `Strangler`, you can put the Rustified service in front of the service you want
to migrate, and out of the box, almost everything should still work.
While migrating you slowly add more logic/routes to the Rust service and automatically those routes
won't be handled by the service you're migrating away from.

## Example

```rust
#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
    // We want to forward requests we don't handle ourselves to localhost:3333
    let strangler_svc = axum_strangler::Strangler::new(
        axum::http::uri::Authority::from_static("127.0.0.1:3333"),
    );

    // Add it as a fallback so everything that isn't covered by routes, get forwarded to the strangled service.
    let router = axum::Router::new()
        .route("/hello",  get(|| async { "Hello, World!" }))
        .fallback_service(strangler_svc);

    axum::Server::bind(&"127.0.0.1:0".parse()?)
        .serve(router.into_make_service())
        .await?;
    Ok(())
}
```

## Feature flags

### `nested-routers`

If you are using an axum router `nested` inside of another one, the route forwarding will not work as expected, as the path that get's forwarded is only the piece of the path defined in the nested router, not the entire path
(so if you nest a router at `/api`, all requests would be forwarded without the `/api`, and only what comes after it).

To fix this, you can enable the feature flag `nested-routers`, which will allow it to retrieve the `OriginalUri` to properly forward it.

### `https`

Allows forwarding to another server that only accepts HTTPS traffic:

```rust
let strangler_svc = axum_strangler::Strangler::builder(
    axum::http::uri::Authority::from_static("127.0.0.1:3333"),
).with_http_scheme(axum_strangler::HttpScheme::HTTPS).build();
```

### `websocket`

Allows the strangler service to also handle websockets, forwarding every message from requester to strangled service and vice versa.

#### TLS

In order to work with websockets over TLS (`wss://`), you'll need to enable additional features.
You can choose which `tokio-tungstenite` dependency you use for tls, all of the three following features map on the counterpart there, but all three enable the `wss://` protocol:

- `websocket-native-tls`
- `websocket-rustls-tls-native-roots`
- `websocket-rustls-tls-webpki-roots`

### `tracing-opentelemetry-text-map-propagation`

Causes the Strangler to propagate tracing information to the stranglee. This could be useful to gather information about what exactly is going on.
This only works if there is an active `opentelemetry` context in the current `tracing` span, and you've installed the `opentelemetry::sdk::propagation::TraceContextPropagator` as the `opentelemetry::global::set_text_map_propagator`.

## Contribution

Unless you explicitly state otherwise, any contribution intentionally submitted
for inclusion in the work by you, as defined in the Apache-2.0 license, shall be
dual licensed as above, without any additional terms or conditions.


CHANGELOG.md
# Changelog

## 0.4.0-rc.5

- Remove `axum` from websocket implementation

## 0.4.0-rc.4

- Add feature `nested-routers` to be able to handle nested axum routers.

## 0.4.0-rc.3

- Update docs
- Update for `axum` `0.6.0`
- Port over partially to `axum-core` (except for websockets)

## 0.4.0-rc.2

- Add a function to allow a user to also easily forward requests, if they want to use it behind their own logic on whether or not to forward it.
- Rename `StranglerService` to `Strangler`

## 0.4.0-rc.1

- Make it optional to rewrite the `host` header to the strangled host.
- Https & Wss support has been put behind the appropriate feature flags ([#12](https://github.com/MidasLamb/axum-strangler/pull/12))
- `tokio` dependency has been made optional ([#10](https://github.com/MidasLamb/axum-strangler/pull/10)), thanks to @syphar.
- `websocket` support has been put behind a feature flag.
- Remove unused `reqwest` dependency ([#7](https://github.com/MidasLamb/axum-strangler/pull/7)), thanks to @syphar.

## 0.3.1

- Add support for HTTPS
- Rewrite HOST header such that proxies down the line can correctly forward it (i.e. tls termination)

## 0.3.0

- Support for websocket strangling.
- Added a bit of tracing.

## 0.2.0

- Use `hyper::Client` instead of `reqwest::Client`
- Pass along the headers

## 0.1.0

- Initial release


src/builder.rs
use std::sync::Arc;

#[cfg(feature = "websocket")]
use crate::WebSocketScheme;
use crate::{
    inner::{InnerStrangler, InnerStranglerService},
    HttpScheme, Strangler,
};

pub struct StranglerBuilder {
    authority: http::uri::Authority,
    http_scheme: HttpScheme,
    #[cfg(feature = "websocket")]
    web_socket_scheme: WebSocketScheme,
    rewrite_strangled_request_host_header: bool,
}

impl StranglerBuilder {
    /// Creates a new strangler with the default options.
    pub fn new(authority: http::uri::Authority) -> Self {
        Self {
            authority,
            http_scheme: HttpScheme::HTTP,
            #[cfg(feature = "websocket")]
            web_socket_scheme: WebSocketScheme::WS,
            rewrite_strangled_request_host_header: false,
        }
    }

    /// The default is `HttpScheme::HTTP`
    pub fn with_http_scheme(self, http_scheme: HttpScheme) -> Self {
        Self {
            http_scheme,
            ..self
        }
    }

    /// The default is `WebSocketScheme::WS`
    #[cfg(feature = "websocket")]
    pub fn with_web_socket_scheme(self, web_socket_scheme: WebSocketScheme) -> Self {
        Self {
            web_socket_scheme,
            ..self
        }
    }

    /// Whether the service should rewrite the `host` header to the strangled target, or leave it be as is.
    /// If the other service is behind e.g. a `traefik` ingress controller or `nginx`, you probably want to set this
    /// to `true`, otherwise those proxies won't know where to send the request.
    pub fn rewrite_strangled_request_host_header(
        self,
        rewrite_strangled_request_host_header: bool,
    ) -> Self {
        Self {
            rewrite_strangled_request_host_header,
            ..self
        }
    }

    /// Turns the builder into a `Strangler`.
    pub fn build(self) -> Strangler {
        let inner: Arc<dyn InnerStrangler + Send + Sync> = match self.http_scheme {
            HttpScheme::HTTP => {
                let inner = InnerStranglerService::new(
                    self.authority,
                    self.http_scheme,
                    #[cfg(feature = "websocket")]
                    self.web_socket_scheme,
                    hyper::Client::new(),
                    self.rewrite_strangled_request_host_header,
                );
                Arc::new(inner)
            }
            #[cfg(feature = "https")]
            HttpScheme::HTTPS => {
                let https = hyper_tls::HttpsConnector::new();
                let client = hyper::Client::builder().build::<_, hyper::Body>(https);
                let inner = InnerStranglerService::new(
                    self.authority,
                    self.http_scheme,
                    #[cfg(feature = "websocket")]
                    self.web_socket_scheme,
                    client,
                    self.rewrite_strangled_request_host_header,
                );
                Arc::new(inner)
            }
        };

        Strangler { inner }
    }
}


src/inner/mod.rs
use async_trait::async_trait;
use http::Uri;

use crate::HttpScheme;

#[cfg(feature = "websocket")]
use crate::WebSocketScheme;

#[cfg(feature = "websocket")]
mod websocket;

#[cfg(feature = "tracing-opentelemetry-text-map-propagation")]
mod tracing_opentelemetry_text_map_propagation;

#[async_trait]
pub(crate) trait InnerStrangler {
    async fn forward_call_to_strangled(
        &self,
        req: http::Request<hyper::body::Body>,
    ) -> axum_core::response::Response;
}

#[async_trait]
impl<C> InnerStrangler for InnerStranglerService<C>
where
    C: hyper::client::connect::Connect + Clone + Send + Sync + 'static,
{
    async fn forward_call_to_strangled(
        &self,
        req: http::Request<hyper::body::Body>,
    ) -> axum_core::response::Response {
        let req = match self.handle_websocket_upgrade_request(req).await {
            Ok(r) => {
                return r;
            }
            Err(r) => r,
        };

        let strangled_authority = self.strangled_authority.clone();
        let strangled_scheme = self.get_http_scheme();

        let (mut req, original_uri) = get_original_uri(req).await;

        let uri = Uri::builder()
            .scheme(strangled_scheme)
            .authority(strangled_authority)
            .path_and_query(original_uri.path_and_query().cloned().unwrap())
            .build()
            .unwrap();

        if self.rewrite_strangled_request_host_header {
            if let Some(host) = req.headers_mut().get_mut("host") {
                *host = http::HeaderValue::from_str(uri.authority().unwrap().as_str()).unwrap()
            }
        }

        #[cfg(feature = "tracing-opentelemetry-text-map-propagation")]
        {
            req =
                tracing_opentelemetry_text_map_propagation::inject_opentelemetry_context_into_request(
                    req,
                );
        }

        *req.uri_mut() = uri;

        let r = self.http_client.request(req).await.unwrap();

        let mut response_builder = axum_core::response::Response::builder();
        response_builder = response_builder.status(r.status());

        if let Some(headers) = response_builder.headers_mut() {
            *headers = r.headers().clone();
        }

        let response = response_builder
            .body(axum_core::body::boxed(r))
            .map_err(|_| http::StatusCode::INTERNAL_SERVER_ERROR);

        match response {
            Ok(response) => response,
            Err(_) => todo!(),
        }
    }
}

#[cfg(not(feature = "nested-routers"))]
async fn get_original_uri(
    req: http::Request<hyper::body::Body>,
) -> (http::Request<hyper::body::Body>, http::Uri) {
    let uri = req.uri().clone();
    (req, uri)
}

#[cfg(feature = "nested-routers")]
async fn get_original_uri(
    req: http::Request<hyper::body::Body>,
) -> (http::Request<hyper::body::Body>, http::Uri) {
    use axum::extract::FromRequestParts;

    let (mut parts, body) = req.into_parts();

    let original_uri = axum::extract::OriginalUri::from_request_parts(&mut parts, &())
        .await
        .unwrap()
        .0;

    let req = http::Request::from_parts(parts, body);

    (req, original_uri)
}

pub(crate) struct InnerStranglerService<C> {
    strangled_authority: http::uri::Authority,
    strangled_http_scheme: HttpScheme,
    #[cfg(feature = "websocket")]
    strangled_web_socket_scheme: WebSocketScheme,
    http_client: hyper::Client<C>,
    rewrite_strangled_request_host_header: bool,
}

impl<C> InnerStranglerService<C>
where
    C: hyper::client::connect::Connect + Clone + Send + Sync + 'static,
{
    pub(crate) fn new(
        strangled_authority: http::uri::Authority,
        strangled_http_scheme: HttpScheme,
        #[cfg(feature = "websocket")] strangled_web_socket_scheme: WebSocketScheme,
        http_client: hyper::Client<C>,
        rewrite_strangled_request_host_header: bool,
    ) -> Self {
        Self {
            strangled_authority,
            strangled_http_scheme,
            #[cfg(feature = "websocket")]
            strangled_web_socket_scheme,
            http_client,
            rewrite_strangled_request_host_header,
        }
    }

    #[cfg(not(feature = "websocket"))]
    async fn handle_websocket_upgrade_request(
        &self,
        req: http::Request<hyper::body::Body>,
    ) -> Result<axum_core::response::Response, http::Request<hyper::body::Body>> {
        Err(req)
    }

    fn get_http_scheme(&self) -> http::uri::Scheme {
        match self.strangled_http_scheme {
            HttpScheme::HTTP => http::uri::Scheme::HTTP,
            #[cfg(feature = "https")]
            HttpScheme::HTTPS => http::uri::Scheme::HTTPS,
        }
    }
}

#[cfg(test)]
mod tests {
    use wiremock::{
        matchers::{header, method, path},
        Mock, MockServer, ResponseTemplate,
    };

    use super::*;

    #[tokio::test]
    async fn no_header_rewriting() {
        let mock_server = MockServer::start().await;

        Mock::given(method("GET"))
            .and(path("/hello"))
            .and(header("host", "something.com"))
            .respond_with(ResponseTemplate::new(200))
            .mount(&mock_server)
            .await;

        let authority = axum::http::uri::Authority::try_from(format!(
            "127.0.0.1:{}",
            mock_server.address().port()
        ))
        .unwrap();

        let client = hyper::client::Client::new();
        let inner = InnerStranglerService::new(
            authority,
            HttpScheme::HTTP,
            #[cfg(feature = "websocket")]
            crate::WebSocketScheme::WS,
            client,
            false,
        );
        let mut request_builder = axum::http::Request::builder()
            .method("GET")
            .uri("http://something.com/hello");
        request_builder.headers_mut().unwrap().insert(
            "host",
            axum::http::HeaderValue::from_static("something.com"),
        );

        let response = inner
            .forward_call_to_strangled(request_builder.body(axum::body::Body::empty()).unwrap())
            .await;

        assert_eq!(response.status(), axum::http::status::StatusCode::OK)
    }

    #[tokio::test]
    async fn header_rewriting() {
        let mock_server = MockServer::start().await;

        let authority = axum::http::uri::Authority::try_from(format!(
            "127.0.0.1:{}",
            mock_server.address().port()
        ))
        .unwrap();

        Mock::given(method("GET"))
            .and(path("/hello"))
            .and(header("host", authority.as_str()))
            .respond_with(ResponseTemplate::new(200))
            .mount(&mock_server)
            .await;

        let client = hyper::client::Client::new();
        let inner = InnerStranglerService::new(
            authority,
            HttpScheme::HTTP,
            #[cfg(feature = "websocket")]
            crate::WebSocketScheme::WS,
            client,
            true,
        );
        let mut request_builder = axum::http::Request::builder()
            .method("GET")
            .uri("http://something.com/hello");
        request_builder.headers_mut().unwrap().insert(
            "host",
            axum::http::HeaderValue::from_static("something.com"),
        );

        let response = inner
            .forward_call_to_strangled(request_builder.body(axum::body::Body::empty()).unwrap())
            .await;

        assert_eq!(response.status(), axum::http::status::StatusCode::OK)
    }
}


examples/simple_server.rs
use axum::{
    extract::ws::{WebSocket, WebSocketUpgrade},
    routing::get,
    Router,
};
use axum_strangler::Strangler;
use tracing_subscriber::prelude::*;

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
    tracing_subscriber::registry()
        .with(
            tracing_subscriber::fmt::layer()
                .with_span_events(tracing_subscriber::fmt::format::FmtSpan::ACTIVE),
        )
        .with(tracing_subscriber::EnvFilter::new(
            "warn,axum_strangler=trace",
        ))
        .init();

    let stranglee_tcp = std::net::TcpListener::bind(format!(
        "127.0.0.1:{}",
        std::env::var("STRANGLEE_PORT").unwrap_or("0".to_owned())
    ))?;
    let stranglee_port = stranglee_tcp.local_addr()?.port();

    println!("Stranglee is listening on port {}", stranglee_port);

    let strangler_tcp = std::net::TcpListener::bind(format!(
        "127.0.0.1:{}",
        std::env::var("STRANGLER_PORT").unwrap_or("0".to_owned())
    ))?;
    let strangler_port = strangler_tcp.local_addr()?.port();

    println!("Strangler is listening on port {}", strangler_port);

    let strangler_svc = Strangler::new(axum::http::uri::Authority::try_from(format!(
        "127.0.0.1:{}",
        stranglee_port
    ))?);

    tokio::spawn(async move {
        let router = Router::new()
            .route("/api/text", get(|| async move { "I'm being strangled" }))
            .route(
                "/api/json",
                get(|| async move { axum::response::Json("A json string") }),
            )
            .route(
                "/api/websocket",
                get(|ws: WebSocketUpgrade| async move { ws.on_upgrade(handle_socket) }),
            );

        axum::Server::from_tcp(stranglee_tcp)
            .unwrap()
            .serve(router.into_make_service())
            .await
            .unwrap();
    });

    let router = Router::new().fallback_service(strangler_svc);
    axum::Server::from_tcp(strangler_tcp)?
        .serve(router.into_make_service())
        .with_graceful_shutdown(async {
            tokio::signal::ctrl_c()
                .await
                .expect("failed to install CTRL+C signal handler");
        })
        .await?;

    Ok(())
}

async fn handle_socket(mut socket: WebSocket) {
    while let Some(msg) = socket.recv().await {
        let msg = if let Ok(msg) = msg {
            msg
        } else {
            // client disconnected
            return;
        };

        if socket.send(msg).await.is_err() {
            // client disconnected
            return;
        }
    }
}


src/inner/tracing_opentelemetry_text_map_propagation.rs
use std::str::FromStr;

use http::HeaderValue;
use opentelemetry::propagation::Injector;
use tracing_opentelemetry::OpenTelemetrySpanExt;

type Request = http::Request<hyper::body::Body>;

pub fn inject_opentelemetry_context_into_request(mut request: Request) -> Request {
    let context = tracing::Span::current().context();

    opentelemetry::global::get_text_map_propagator(|injector| {
        injector.inject_context(&context, &mut RequestCarrier::new(&mut request))
    });

    request
}

// Adapted from reqwest-middleware::otel
struct RequestCarrier<'a> {
    request: &'a mut Request,
}

impl<'a> RequestCarrier<'a> {
    pub fn new(request: &'a mut Request) -> Self {
        RequestCarrier { request }
    }
}

impl<'a> Injector for RequestCarrier<'a> {
    fn set(&mut self, key: &str, value: String) {
        let header_name = hyper::header::HeaderName::from_str(key).expect("Must be header name");
        let header_value = HeaderValue::from_str(&value).expect("Must be a header value");
        self.request.headers_mut().insert(header_name, header_value);
    }
}


src/lib.rs
//! # Axum Strangler
//! A `tower_service::Service` for use in the `axum` web framework to apply the  Strangler Fig pattern.
//! This makes "strangling" a bit easier, as everything that is handled by the "strangler" will
//! automatically no longer be forwarded to the "stranglee" or "strangled application" (a.k.a. the old application).
//!
//! ## Example
//! ```rust
//! #[tokio::main]
//! async fn main() -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
//!     let strangler = axum_strangler::Strangler::new(
//!         axum::http::uri::Authority::from_static("127.0.0.1:3333"),
//!     );
//!     let router = axum::Router::new().fallback_service(strangler);
//!     axum::Server::bind(&"127.0.0.1:0".parse()?)
//!         .serve(router.into_make_service())
//!         # .with_graceful_shutdown(async {
//!         # // Shut down immediately
//!         # })
//!         .await?;
//!     Ok(())
//! }
//! ```
//!
//! ## Caveats
//! Note that when registering a route with `axum`, all requests will be handled by it, even if you don't register anything for the specific method.
//! This means that in the following snippet, requests for `/new` with the method
//! POST, PUT, DELETE, OPTIONS, HEAD, PATCH, or TRACE will no longer be forwarded to the strangled application:
//! ```rust
//! async fn handler() {}
//!
//! #[tokio::main]
//! async fn main() -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
//!     let strangler = axum_strangler::Strangler::new(
//!         axum::http::uri::Authority::from_static("127.0.0.1:3333"),
//!     );
//!     let router = axum::Router::new()
//!         .route(
//!             "/test",
//!              axum::routing::get(handler)
//!         )
//!         .fallback_service(strangler);
//!     axum::Server::bind(&"127.0.0.1:0".parse()?)
//!         .serve(router.into_make_service())
//!         # .with_graceful_shutdown(async {
//!         # // Shut down immediately
//!         # })
//!         .await?;
//!     Ok(())
//! }
//! ```
//!
//! If you only want to implement a single method and still forward the rest, you can do so by adding the strangler as the fallback
//! for that specific `MethodRouter`:
//! ```rust
//! async fn handler() {}
//!
//! #[tokio::main]
//! async fn main() -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
//!     let strangler = axum_strangler::Strangler::new(
//!         axum::http::uri::Authority::from_static("127.0.0.1:3333"),
//!     );
//!     let router = axum::Router::new()
//!         .route(
//!             "/test",
//!             axum::routing::get(handler)
//!                 .fallback_service(strangler.clone())
//!         )
//!         .fallback_service(strangler);
//!     axum::Server::bind(&"127.0.0.1:0".parse()?)
//!         .serve(router.into_make_service())
//!         # .with_graceful_shutdown(async {
//!         # // Shut down immediately
//!         # })
//!         .await?;
//!     Ok(())
//! }
//! ```
//!
//! ## Websocket support
//! If you enable the feature `websocket` (and possibly one of the supporting tls ones: websocket-native-tls,
//! websocket-rustls-tls-native-roots, websocket-rustls-tls-webpki-roots), a websocket will be set up, and each websocket
//! message will be relayed.
//!
//! ## Tracing propagation
//! Enabling the `tracing-opentelemetry-text-map-propagation` feature, will cause traceparent header to be set on
//! requests that get forwarded, based on the current `tracing` (& `tracing-opentelemetry`) context.
//!
//! Note that this requires the `opentelemetry` `TextMapPropagator` to be installed.

#![cfg_attr(docsrs, feature(doc_cfg))]

use std::{
    convert::Infallible,
    future::Future,
    pin::Pin,
    sync::Arc,
    task::{Context, Poll},
};

use tower_service::Service;

mod builder;
mod inner;

pub enum HttpScheme {
    HTTP,
    #[cfg(any(docsrs, feature = "https"))]
    #[cfg_attr(docsrs, doc(cfg(feature = "https")))]
    HTTPS,
}

#[cfg(any(docsrs, feature = "websocket"))]
#[cfg_attr(docsrs, doc(cfg(feature = "websocket")))]
pub enum WebSocketScheme {
    WS,
    #[cfg(any(
        feature = "websocket-native-tls",
        feature = "websocket-rustls-tls-native-roots",
        feature = "websocket-rustls-tls-webpki-roots"
    ))]
    #[cfg_attr(
        docsrs,
        doc(cfg(any(
            feature = "websocket-native-tls",
            feature = "websocket-rustls-tls-native-roots",
            feature = "websocket-rustls-tls-webpki-roots"
        )))
    )]
    WSS,
}

/// Forwards all requests to another application.
/// Can be used in a lot of places, but the most common one would be as a `.fallback` on an `axum` `Router`.
/// # Example
/// ```rust
/// #[tokio::main]
/// async fn main() -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
///     let strangler_svc = axum_strangler::Strangler::new(
///         axum::http::uri::Authority::from_static("127.0.0.1:3333"),
///     );
///     let router = axum::Router::new().fallback_service(strangler_svc);
///     axum::Server::bind(&"127.0.0.1:0".parse()?)
///         .serve(router.into_make_service())
///         # .with_graceful_shutdown(async {
///         # // Shut down immediately
///         # })
///         .await?;
///     Ok(())
/// }
/// ```
#[derive(Clone)]
pub struct Strangler {
    inner: Arc<dyn inner::InnerStrangler + Send + Sync>,
}

impl Strangler {
    /// Creates a new `Strangler` with the default options.
    /// For more control, see [`builder::StranglerBuilder`]
    pub fn new(strangled_authority: http::uri::Authority) -> Self {
        Strangler::builder(strangled_authority).build()
    }

    pub fn builder(strangled_authority: http::uri::Authority) -> builder::StranglerBuilder {
        builder::StranglerBuilder::new(strangled_authority)
    }

    /// Forwards the request to the strangled service.
    /// Meant to be used when you want to send something to the strangled application
    /// based on some custom logic.
    pub async fn forward_to_strangled(
        &self,
        req: http::Request<hyper::body::Body>,
    ) -> axum_core::response::Response {
        self.inner.forward_call_to_strangled(req).await
    }
}

impl Service<http::Request<hyper::body::Body>> for Strangler {
    type Response = axum_core::response::Response;
    type Error = Infallible;
    type Future = Pin<Box<dyn Future<Output = Result<Self::Response, Self::Error>> + Send>>;

    fn poll_ready(&mut self, _cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
        Poll::Ready(Ok(()))
    }

    fn call(&mut self, req: http::Request<hyper::body::Body>) -> Self::Future {
        let inner = self.inner.clone();

        let fut = async move { Ok(inner.forward_call_to_strangled(req).await) };
        Box::pin(fut)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use axum::{body::HttpBody, Router};
    use wiremock::{
        matchers::{method, path},
        Mock, ResponseTemplate,
    };

    /// Create a mock service that's not connecting to anything.
    fn make_svc() -> Strangler {
        Strangler::new(axum::http::uri::Authority::from_static("127.0.0.1:0"))
    }

    #[tokio::test]
    async fn can_be_used_as_fallback() {
        let router = Router::new().fallback_service(make_svc());
        axum::Server::bind(&"0.0.0.0:0".parse().unwrap()).serve(router.into_make_service());
    }

    #[tokio::test]
    async fn can_be_used_for_a_route() {
        let router = Router::new().route_service("/api", make_svc());
        axum::Server::bind(&"0.0.0.0:0".parse().unwrap()).serve(router.into_make_service());
    }

    #[tokio::test]
    async fn proxies_strangled_http_service() {
        let mock_server = wiremock::MockServer::start().await;

        Mock::given(method("GET"))
            .and(path("/api/something"))
            .respond_with(ResponseTemplate::new(200).set_body_string("I'm being strangled"))
            .mount(&mock_server)
            .await;

        let strangler_svc = Strangler::new(
            axum::http::uri::Authority::try_from(format!(
                "127.0.0.1:{}",
                mock_server.address().port()
            ))
            .unwrap(),
        );

        let router = Router::new().fallback_service(strangler_svc);

        let req = http::Request::get("/api/something")
            .body(hyper::body::Body::empty())
            .unwrap();
        let mut res = router.clone().call(req).await.unwrap();

        assert_eq!(res.status(), http::StatusCode::OK);

        assert_eq!(
            res.body_mut().data().await.unwrap().unwrap(),
            "I'm being strangled".as_bytes()
        );
    }

    #[cfg(feature = "nested-routers")]
    #[tokio::test]
    async fn handles_nested_routers() {
        let mock_server = wiremock::MockServer::start().await;

        Mock::given(method("GET"))
            .and(path("/api/something"))
            .respond_with(ResponseTemplate::new(200))
            .mount(&mock_server)
            .await;

        Mock::given(method("GET"))
            .and(path("/api/something-else"))
            .respond_with(ResponseTemplate::new(418))
            .mount(&mock_server)
            .await;

        let strangler_svc = Strangler::new(
            axum::http::uri::Authority::try_from(format!(
                "127.0.0.1:{}",
                mock_server.address().port()
            ))
            .unwrap(),
        );

        let nested_router = Router::new()
            .route_service("/something", strangler_svc.clone())
            .route_service("/something-else", strangler_svc.clone());
        let router = Router::new()
            .nest("/api", nested_router)
            .fallback_service(strangler_svc);

        let req = http::Request::get("/api/something")
            .body(hyper::body::Body::empty())
            .unwrap();
        let res = router.clone().call(req).await.unwrap();
        assert_eq!(res.status(), http::StatusCode::OK);

        let req = http::Request::get("/api/something-else")
            .body(hyper::body::Body::empty())
            .unwrap();
        let res = router.clone().call(req).await.unwrap();
        assert_eq!(res.status(), http::StatusCode::IM_A_TEAPOT);

        let req = http::Request::get("/not-api/something-else")
            .body(hyper::body::Body::empty())
            .unwrap();
        let res = router.clone().call(req).await.unwrap();
        assert_eq!(res.status(), http::StatusCode::NOT_FOUND);
    }
}
